[CI:BUILD] Packit: add jobs for downstream Fedora package builds

lsm5 opened the pull request:
Get rid of `podman.spec.rpkg` in favour of
`rpm/podman.spec` which gets synced with fedora dist-git on every upstream release. The version in the new spec file is set to `0` by default and gets updated by packit automatically on every packit task.

For local manual rpm builds using the spec, the helper script in the `rpm/` subdir will update the Version field with the latest version found in the upstream repo.

Packit will automatically create a PR on fedora dist-git on every new upstream release. A sample PR will look like:
https://src.fedoraproject.org/rpms/container-selinux/pull-request/10#

A dry run for this can be triggered using:
`$ packit propose-downstream --local-content`

To run this command locally, you would need to have your packit user-configuration-file set.
Ref: https://packit.dev/docs/configuration/#user-configuration-file

along with a fedora api key created at:
https://src.fedoraproject.org/settings#nav-api-tab with sufficient ACLs.

Also includes a revised `package` Makefile target which will build rpms
using `rpm/podman.spec`. Fixes: #18421.

[NO NEW TESTS NEEDED]

<!--
Thanks for sending a pull request!

Please make sure you've read our contributing guidelines and how to submit a pull request (https://github.com/containers/podman/blob/main/CONTRIBUTING.md#submitting-pull-requests).

In case you're only changing docs, make sure to prefix the pull-request title with "[CI:DOCS]". That will prevent functional tests from running and save time and energy.

Finally, be sure to sign commits with your real name. Since by opening
a PR you already have commits, you can add signatures if needed with
something like `git commit -s --amend`.
-->

#### Does this PR introduce a user-facing change?

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes, please follow the Kubernetes model:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
None
```

lsm5 commented:
@mheon any idea if sqlite3 doesn't play well with centos 8 stream?

```
# github.com/containers/podman/vendor/github.com/mattn/go-sqlite3
_build/src/github.com/containers/podman/vendor/github.com/mattn/go-sqlite3/sqlite3.go:74:1: warning: '_sqlite3_exec' defined but not used [-Wunused-function]
 _sqlite3_exec(sqlite3* db, const char* pcmd, long long* rowid, long long* changes)
 ^~~~~~~~~~~~~
error: Bad exit status from /var/tmp/rpm-tmp.k4tlck (%build)
```

works on centos 9 stream and fedora.

lsm5 commented:
@mheon ignore me, the actual error was `btrfs/ioctl.h: no such file or directory` way up in the build log.

lsm5 commented:
@rhatdan @edsantiago @mheon @containers/podman-maintainers PTAL . Would like this in v4.5 before we cut v4.5.1.

@Steve973 This should make `make package` work for you on rhel8. Feel free to checkout the PR commit and give it a try.

lsm5 commented:
err, I need to handle containers-common dependency for rhel. Hold off on review.

mheon commented:
LGTM on my end, but someone more familiar with RPMs should review before merge

edsantiago commented:
@lsm5 help please, I'm feeling very dumb. I'm trying to figure out where the new specfile comes from, so I can look at the minimalest set of diffs. I've compared against [fedpkg](https://src.fedoraproject.org/rpms/podman) @ main and f38 and f37, and those diffs are huge. Where can I find the smallest set of diffs?

Also, when you say "gets synced with fedora dist-git", do you mean that Packit pushes to fedpkg, or that Packit pulls from fedpkg, or something else? I'm guessing Packit pushes something to fedpkg, but am not sure what that looks like (`rpm/podman.spec` with updated NVR???)

I realize these are stupid questions, and apologize.

lsm5 commented:
@edsantiago if  it makes things easier, I can first push the `rpm/podman.spec` separately to dist-git rawhide, so that your new diff will be minimal. wdyt ?

edsantiago commented:
Thanks, but I'm good. Your explanation helped. Am in the middle of reviewing now.

edsantiago commented on rpm/podman.spec:
```diff
@@ -0,0 +1,418 @@
+%gobuild -o bin/rootlessport %{import_path}/cmd/rootlessport
+
+export BASEBUILDTAGS="seccomp exclude_graphdriver_devicemapper $(hack/systemd_tag.sh) $(hack/libsubid_tag.sh)"
+export BUILDTAGS="$BASEBUILDTAGS $(hack/btrfs_installed_tag.sh) $(hack/btrfs_tag.sh)"
```
This lost `hack/libdm_tag.sh`. Can someone confirm whether or not that's OK?

lsm5 replied:
err, nope. I was mistaken. The libdm_tag has other tags and not exclude_graphdriver_devicemapper. Let me double-check on this.

lsm5 replied:
@edsantiago you're right. It was a mistake to remove that tag. I'll add it back.

edsantiago commented on rpm/update-spec-provides.sh:
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
+
+# This script will update the Version field in the spec which is set to 0 by
+# default. Useful for local manual rpm builds where the Version needs to be set
+# correctly.
```
Comment is copypasted from the next script, and is misleading. And actually, since this script isn't even used, I might suggest removing it and adding it back when you fix the TODO. I won't block over that though.

lsm5 replied:
err, yes I need to change that comment. I'm actually also checking with the packit people on the side.

lsm5 replied:
updated comment.

edsantiago commented on rpm/update-spec-version.sh:
```diff
@@ -0,0 +1,11 @@
+# default. Useful for local manual rpm builds where the Version needs to be set
+# correctly.
+
+SPEC_FILE=$(pwd)/podman.spec
```
Is this really `$(pwd)`? I looked in `.packit.yml` and don't see where it cd's to the `rpm` directory?

I would rather see something like `$(dirname $0)/podman.spec`, it seems safer. But it's very likely that I'm missing a key piece of Packit magic.

lsm5 replied:
`update-spec-version.sh` is only if someone needs a manual build or use `make package` locally. Doesn't get used by packit. I can update the comment to make that clear.

edsantiago reviewed (commented):
First pass done. Some concerns, some suggestions, but if this is working right now I'm fine with merging and then adding a followup PR. I'm reluctant to re-review this whole thing again :-)

lsm5 commented on rpm/podman.spec:
```diff
@@ -0,0 +1,418 @@
+# Keep Version in upstream specfile at 0. It will be automatically set
+# to the correct value by Packit for copr and koji builds.
+# IGNORE this comment if you're looking at it in dist-git.
+Version: 0
```
hmm, yes. I think I can stick that in the build section. Thanks for the tip. Let me give it a try.

lsm5 replied:
i'll leave this for the future. It's pretty much impossible for rpmbuild to proceed with the current spec file when Version is 0, so this version check isn't an urgent requirement.

lsm5 commented on rpm/podman.spec:
```diff
@@ -0,0 +1,418 @@
+%else
+Epoch: 5
+%endif
+# Keep Version in upstream specfile at 0. It will be automatically set
+# to the correct value by Packit for copr and koji builds.
+# IGNORE this comment if you're looking at it in dist-git.
```
Sure, I can update the comment. Thanks

edsantiago commented:
Up to you. I'm comfortable using `git range-diff` for small to medium changesets.

lsm5 commented:
updated PR with libdm_tag re-added and better comments in spec and the update-spec-provides script. Also check https://src.fedoraproject.org/rpms/podman/pull-request/66#request_diff for a sample of the dist-git diff that would be created when we do a new release. 

Btw, I've removed the `selinux` build tag because it's not needed anymore: https://github.com/opencontainers/selinux/blob/main/README.md#usage

lsm5 commented:
@edsantiago PTAL. If all the major comments are addressed, let's get this in. Thanks!

lsm5 commented:
Btw, I've submitted the spec file change to dist-git rawhide branch already, in case you wanna run another diff between spec files.

edsantiago commented:
`git range-diff` shows (most of) what I expected to see. The rest of my concerns can be addressed later (please don't forget :-)

/lgtm

lsm5 commented:
/cherrypick v4.5
